client-spark/shuffle-manager-common/src/main/java/org/apache/spark/shuffle/rss/ShuffleInMemorySorter.java

```diff
@@ -57,40 +58,47 @@ private int getUsableCapacity() {
     return (int) (array.size() / 2);
   }
 
-  public void free() {
-    if (array != null) {
-      consumer.freeArray(array);
-      array = null;
-    }
+  public long getInitialSize() {
+    return initialSize;
   }
 
-  public void reset() {
-    // Reset `pos` here so that `spill` triggered by the below `allocateArray` will be no-op.
-    pos = 0;
-    if (consumer != null && array != null) {
-
-      consumer.freeArray(array);
-      // As `array` has been released, we should set it to  `null` to avoid accessing it before
-      // `allocateArray` returns. `usableCapacity` is also set to `0` to avoid any codes writing
-      // data to `ShuffleInMemorySorter` when `array` is `null` (e.g., in
-      // ShuffleExternalSorter.growPointerArrayIfNecessary, we may try to access
-      // `ShuffleInMemorySorter` when `allocateArray` throws SparkOutOfMemoryError).
+  public void freeMemory() {
+    if (consumer != null) {
+      if (array != null) {
+        consumer.freeArray(array);
+      }
+
+      // Set the array to null instead of allocating a new array. Allocating an array could have
+      // triggered another spill and this method already is called from UnsafeExternalSorter when
+      // spilling. Attempting to allocate while spilling is dangerous, as we could be holding onto
+      // a large partially complete allocation, which may prevent other memory from being allocated.
+      // Instead we will allocate the new array when it is necessary.
       array = null;
       usableCapacity = 0;
-      array = consumer.allocateArray(initialSize);
-      usableCapacity = getUsableCapacity();
     }
+    pos = 0;
+  }
+
+  /** @return the number of records that have been inserted into this sorter. */
+  public int numRecords() {
+    return pos / 2;
```

Comment:
In spark, when call `insertRecord`,  `pos++` will be called twice. But in RSS we only called once, here should be 
```
return pos;
```